Error while checking build/O2/fullCI for 10ada2806455d280689a7a4aaeae88e1f91156f7 at 2022-10-25 11:58:
```
## sw/BUILD/o2codechecker-latest/log
100% tests passed, 0 tests failed out of 1


## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
SOURCES/O2/10151-slc8_x86-64/0/Detectors/ITSMFT/ITS/base/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Detectors/ITSMFT/common/base/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Detectors/ITSMFT/ITS/tracking/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/DataFormats/Detectors/ITSMFT/common/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Detectors/ITSMFT/ITS/reconstruction/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Detectors/ITSMFT/common/reconstruction/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Framework/Utils/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/DataFormats/Detectors/ITSMFT/ITS/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Steer/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Generators/include" "-I" "/sw/SOURCES/O2/10151-slc8_x86-64/0/Detectors/ITSMFT/ITS/tracking/GPU" "-D" "NDEBUG" "-internal-isystem" "/usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8" "-internal-isystem" "/usr/lib/gcc/x86/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10151-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10151/10ada2806455d280689a7a4aaeae88e1f91156f7/build_O2_fullCI/pretty.html).

--- a/Detectors/ITSMFT/MFT/tracking/include/MFTTracking/IOUtils.h
+++ b/Detectors/ITSMFT/MFT/tracking/include/MFTTracking/IOUtils.h
@@ -61,6 +61,12 @@ int loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& events, gsl::s
                     gsl::span<const unsigned char>::iterator& pattIt, const itsmft::TopologyDictionary* dict,
                     const dataformats::MCTruthContainer<MCCompLabel>* mClsLabels = nullptr, const o2::mft::Tracker<T>* tracker = nullptr);
 
+template <typename T>
+int loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& events, gsl::span<const itsmft::CompClusterExt> clusters,
+                    gsl::span<const unsigned char>::iterator& pattIt, const itsmft::TopologyDictionary* dict,
+                    const dataformats::MCTruthContainer<MCCompLabel>* mClsLabels, const o2::mft::Tracker<T>* tracker,
+                    ROFFilter& filter);
+
 void convertCompactClusters(gsl::span<const itsmft::CompClusterExt> clusters,
                             gsl::span<const unsigned char>::iterator& pattIt,
                             std::vector<o2::BaseCluster<float>>& output,

--- a/Detectors/ITSMFT/MFT/tracking/include/MFTTracking/MFTTrackingParam.h
+++ b/Detectors/ITSMFT/MFT/tracking/include/MFTTracking/MFTTrackingParam.h
@@ -67,6 +67,14 @@ struct MFTTrackingParam : public o2::conf::ConfigurableParamHelper<MFTTrackingPa
   /// road for CA algo : cylinder or cone (default)
   Bool_t CAConeRadius = kFALSE;
 
+  // cuts to reject to low or too high mult events, or externally provided IRFrames
+  float cutMultClusLow = 0;   /// reject ROF with estimated cluster mult. below this value (no cut if <0)
+  float cutMultClusHigh = -1; /// reject ROF with estimated cluster mult. above this value (no cut if <0)
+  bool irFramesOnly = false;  ///< track only ROFs that overlap one of the IRFrames (provided externally by ITS)
+
+  bool isMultCutRequested() const { return cutMultClusLow >= 0.f && cutMultClusHigh > 0.f; };
+  bool isPassingMultCut(float mult) const { return mult >= cutMultClusLow && (mult <= cutMultClusHigh || cutMultClusHigh <= 0.f); }
+
   O2ParamDef(MFTTrackingParam, "MFTTracking");
 };
 

--- a/Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Tracker.h
+++ b/Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Tracker.h
@@ -24,6 +24,8 @@
 #include "MathUtils/Utils.h"
 #include "MathUtils/Cartesian.h"
 #include "DataFormatsMFT/TrackMFT.h"
+#include "DataFormatsITSMFT/ROFRecord.h"
+#include "CommonDataFormat/IRFrame.h"
 #include "SimulationDataFormat/MCCompLabel.h"
 #include "SimulationDataFormat/MCTruthContainer.h"
 #include "DataFormatsParameters/GRPObject.h"
@@ -32,6 +34,10 @@ namespace o2
 {
 namespace mft
 {
+
+using o2::dataformats::IRFrame;
+using o2::itsmft::ROFRecord;
+typedef std::function<bool(const ROFRecord&)> ROFFilter;
 
 class T;
 
@@ -59,7 +65,6 @@ class Tracker : public TrackerConfig
   void findLTFTracks(ROframe<T>&);
   void findCATracks(ROframe<T>&);
   bool fitTracks(ROframe<T>&);
-
   void computeTracksMClabels(const std::vector<T>&);
 
   void initialize(bool fullClusterScan = false);

--- a/Detectors/ITSMFT/MFT/tracking/src/IOUtils.cxx
+++ b/Detectors/ITSMFT/MFT/tracking/src/IOUtils.cxx
@@ -29,6 +29,7 @@
 #include "MathUtils/Cartesian.h"
 #include "SimulationDataFormat/MCCompLabel.h"
 #include "SimulationDataFormat/MCTruthContainer.h"
+#include "MFTTracking/MFTTrackingParam.h"
 
 namespace o2
 {
@@ -37,7 +38,7 @@ namespace mft
 
 //_________________________________________________________
 template <typename T>
-int ioutils::loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& event, gsl::span<const itsmft::CompClusterExt> clusters, gsl::span<const unsigned char>::iterator& pattIt, const itsmft::TopologyDictionary* dict, const dataformats::MCTruthContainer<MCCompLabel>* mcLabels, const o2::mft::Tracker<T>* tracker)
+int ioutils::loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& event, gsl::span<const itsmft::CompClusterExt> clusters, gsl::span<const unsigned char>::iterator& pattIt, const itsmft::TopologyDictionary* dict, const dataformats::MCTruthContainer<MCCompLabel>* mcLabels, const o2::mft::Tracker<T>* tracker, ROFFilter& filter)
 {
   event.clear();
   GeometryTGeo* geom = GeometryTGeo::Instance();
@@ -45,13 +46,24 @@ int ioutils::loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& event
   int clusterId{0};
   auto first = rof.getFirstEntry();
   auto clusters_in_frame = rof.getROFData(clusters);
-  event.Reserve(clusters_in_frame.size());
+  auto nClusters = clusters_in_frame.size();
+
+  bool skip_ROF = true;
+  auto& trackingParam = MFTTrackingParam::Instance();
+  if (filter(rof) && trackingParam.isPassingMultCut(nClusters)) {
+    LOG(debug) << " ROF selected! ; nClusters = " << nClusters << " ;   orbit = " << rof.getBCData().orbit << " ; bc = " << rof.getBCData().bc;
+    skip_ROF = false;
+    event.Reserve(nClusters);
+  } else {
+    nClusters = 0;
+  }
+
   for (auto& c : clusters_in_frame) {
     auto sensorID = c.getSensorID();
     int layer = geom->getLayer(sensorID);
     auto pattID = c.getPatternID();
     o2::math_utils::Point3D<float> locXYZ;
-    float sigmaX2 = ioutils::DefClusError2Row, sigmaY2 = ioutils::DefClusError2Col; //Dummy COG errors (about half pixel size)
+    float sigmaX2 = ioutils::DefClusError2Row, sigmaY2 = ioutils::DefClusError2Col; // Dummy COG errors (about half pixel size)
     if (pattID != itsmft::CompCluster::InvalidPatternID) {
       sigmaX2 = dict->getErr2X(pattID); // ALPIDE local X coordinate => MFT global X coordinate (ALPIDE rows)
       sigmaY2 = dict->getErr2Z(pattID); // ALPIDE local Z coordinate => MFT global Y coordinate (ALPIDE columns)
@@ -64,6 +76,10 @@ int ioutils::loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& event
     } else {
       o2::itsmft::ClusterPattern patt(pattIt);
       locXYZ = dict->getClusterCoordinates(c, patt, false);
+    }
+    if (skip_ROF) { // Skip filtered-out ROFs after processing pattIt
+      clusterId++;
+      continue;
     }
     // Transformation to the local --> global
     auto gloXYZ = geom->getMatrixL2G(sensorID) * locXYZ;
@@ -82,7 +98,7 @@ int ioutils::loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& event
     event.addClusterExternalIndexToLayer(layer, first + clusterId);
     clusterId++;
   }
-  return clusters_in_frame.size();
+  return nClusters;
 }
 
 //_________________________________________________________
@@ -121,6 +137,16 @@ void ioutils::convertCompactClusters(gsl::span<const itsmft::CompClusterExt> clu
     cl3d.setErrors(sigmaX2, sigmaY2, 0);
   }
 }
+
+//_________________________________________________________
+template <typename T>
+int ioutils::loadROFrameData(const o2::itsmft::ROFRecord& rof, ROframe<T>& event, gsl::span<const itsmft::CompClusterExt> clusters, gsl::span<const unsigned char>::iterator& pattIt, const itsmft::TopologyDictionary* dict, const dataformats::MCTruthContainer<MCCompLabel>* mcLabels, const o2::mft::Tracker<T>* tracker)
+{
+  ROFFilter noFilter = [](const o2::itsmft::ROFRecord r) { return true; };
+  return ioutils::loadROFrameData(rof, event, clusters, pattIt, dict, mcLabels, tracker, noFilter);
+}
+
+//_________________________________________________________
 template int o2::mft::ioutils::loadROFrameData<o2::mft::TrackLTF>(const o2::itsmft::ROFRecord&, ROframe<o2::mft::TrackLTF>&, gsl::span<const itsmft::CompClusterExt>,
                                                                   gsl::span<const unsigned char>::iterator&, const itsmft::TopologyDictionary*,
                                                                   const dataformats::MCTruthContainer<MCCompLabel>*, const o2::mft::Tracker<o2::mft::TrackLTF>*);
@@ -128,6 +154,15 @@ template int o2::mft::ioutils::loadROFrameData<o2::mft::TrackLTF>(const o2::itsm
 template int o2::mft::ioutils::loadROFrameData<o2::mft::TrackLTFL>(const o2::itsmft::ROFRecord&, ROframe<o2::mft::TrackLTFL>&, gsl::span<const itsmft::CompClusterExt>,
                                                                    gsl::span<const unsigned char>::iterator&, const itsmft::TopologyDictionary*,
                                                                    const dataformats::MCTruthContainer<MCCompLabel>*, const o2::mft::Tracker<o2::mft::TrackLTFL>*);
+template int o2::mft::ioutils::loadROFrameData<o2::mft::TrackLTF>(const o2::itsmft::ROFRecord&, ROframe<o2::mft::TrackLTF>&, gsl::span<const itsmft::CompClusterExt>,
+                                                                  gsl::span<const unsigned char>::iterator&, const itsmft::TopologyDictionary*,
+                                                                  const dataformats::MCTruthContainer<MCCompLabel>*, const o2::mft::Tracker<o2::mft::TrackLTF>*,
+                                                                  ROFFilter& filter);
+
+template int o2::mft::ioutils::loadROFrameData<o2::mft::TrackLTFL>(const o2::itsmft::ROFRecord&, ROframe<o2::mft::TrackLTFL>&, gsl::span<const itsmft::CompClusterExt>,
+                                                                   gsl::span<const unsigned char>::iterator&, const itsmft::TopologyDictionary*,
+                                                                   const dataformats::MCTruthContainer<MCCompLabel>*, const o2::mft::Tracker<o2::mft::TrackLTFL>*,
+                                                                   ROFFilter& filter);
 
 } // namespace mft
 } // namespace o2

--- a/Detectors/ITSMFT/MFT/tracking/src/Tracker.cxx
+++ b/Detectors/ITSMFT/MFT/tracking/src/Tracker.cxx
@@ -81,10 +81,27 @@ void Tracker<T>::initConfig(const MFTTrackingParam& trkParam, bool printConfig)
 
   if (printConfig) {
     LOG(info) << "Configurable tracker parameters:";
+    switch (trkParam.trackmodel) {
+      case o2::mft::MFTTrackModel::Helix:
+        LOG(info) << "Fwd track model     = Helix";
+        break;
+      case o2::mft::MFTTrackModel::Quadratic:
+        LOG(info) << "Fwd track model     = Quadratic";
+        break;
+      case o2::mft::MFTTrackModel::Linear:
+        LOG(info) << "Fwd track model     = Linear";
+        break;
+      case o2::mft::MFTTrackModel::Optimized:
+        LOG(info) << "Fwd track model     = Optimized";
+        break;
+    }
+    LOG(info) << "alignResidual       = " << trkParam.alignResidual;
     LOG(info) << "MinTrackPointsLTF   = " << mMinTrackPointsLTF;
     LOG(info) << "MinTrackPointsCA    = " << mMinTrackPointsCA;
     LOG(info) << "MinTrackStationsLTF = " << mMinTrackStationsLTF;
     LOG(info) << "MinTrackStationsCA  = " << mMinTrackStationsCA;
+    LOG(info) << "LTFConeRadius       = " << (trkParam.LTFConeRadius ? "true" : "false");
+    LOG(info) << "CAConeRadius        = " << (trkParam.CAConeRadius ? "true" : "false");
     LOG(info) << "LTFclsRCut          = " << mLTFclsRCut;
     LOG(info) << "ROADclsRCut         = " << mROADclsRCut;
     LOG(info) << "RBins               = " << mRBins;
@@ -93,6 +110,13 @@ void Tracker<T>::initConfig(const MFTTrackingParam& trkParam, bool printConfig)
     LOG(info) << "LTFinterBinWin      = " << mLTFinterBinWin;
     LOG(info) << "FullClusterScan     = " << (trkParam.FullClusterScan ? "true" : "false");
     LOG(info) << "forceZeroField      = " << (trkParam.forceZeroField ? "true" : "false");
+    LOG(info) << "MFTRadLength        = " << trkParam.MFTRadLength;
+    LOG(info) << "irFramesOnly        = " << (trkParam.irFramesOnly ? "true" : "false");
+    LOG(info) << "isMultCutRequested  = " << (trkParam.isMultCutRequested() ? "true" : "false");
+    if (trkParam.isMultCutRequested()) {
+      LOG(info) << "cutMultClusLow      = " << trkParam.cutMultClusLow;
+      LOG(info) << "cutMultClusHigh     = " << trkParam.cutMultClusHigh;
+    }
   }
 }
 

--- a/Detectors/ITSMFT/MFT/workflow/CMakeLists.txt
+++ b/Detectors/ITSMFT/MFT/workflow/CMakeLists.txt
@@ -27,7 +27,8 @@ o2_add_library(MFTWorkflow
                                      O2::MFTAssessment
                                      O2::DataFormatsMFT
                                      O2::ITSMFTWorkflow
-                                     O2::MFTAlignment)
+                                     O2::MFTAlignment
+                                     O2::GlobalTrackingWorkflowReaders)
 o2_add_executable(reco-workflow
                   SOURCES src/mft-reco-workflow.cxx
                   COMPONENT_NAME mft

--- a/Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/TrackerSpec.h
+++ b/Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/TrackerSpec.h
@@ -16,6 +16,7 @@
 
 #include "MFTTracking/Tracker.h"
 #include "DetectorsBase/GRPGeomHelper.h"
+#include "ITSMFTBase/DPLAlpideParam.h"
 
 #include "Framework/DataProcessorSpec.h"
 #include "MFTTracking/TrackCA.h"
@@ -43,7 +44,16 @@ class TrackerDPL : public o2::framework::Task
 
  private:
   void updateTimeDependentParams(framework::ProcessingContext& pc);
+  ///< MFT readout mode
+  bool mMFTTriggered = false; ///< MFT readout is triggered
 
+  ///< set MFT ROFrame duration in microseconds
+  void setMFTROFrameLengthMUS(float fums);
+  ///< set MFT ROFrame duration in BC (continuous mode only)
+  void setMFTROFrameLengthInBC(int nbc);
+  int mMFTROFrameLengthInBC = 0;       ///< MFT RO frame in BC (for MFT cont. mode only)
+  float mMFTROFrameLengthMUS = -1.;    ///< MFT RO frame in \mus
+  float mMFTROFrameLengthMUSInv = -1.; ///< MFT RO frame in \mus inverse
   bool mUseMC = false;
   bool mFieldOn = true;
   int mNThreads = 4;
@@ -67,6 +77,24 @@ class TrackerDPL : public o2::framework::Task
                                                    "FitTracks",
                                                    "ComputeLabels"};
   TStopwatch mTimer[NStopWatches];
+
+  ROFFilter createIRFrameFilter(gsl::span<const o2::dataformats::IRFrame> irframes)
+  {
+    return [this, irframes](const ROFRecord& rof) {
+      InteractionRecord rofStart{rof.getBCData()};
+      InteractionRecord rofEnd = rofStart + mMFTROFrameLengthInBC - 1;
+      IRFrame ref(rofStart, rofEnd);
+      for (const auto& ir : irframes) {
+        if (ir.info > 0) {
+          auto overlap = ref.getOverlap(ir);
+          if (overlap.isValid()) {
+            return true;
+          }
+        }
+      }
+      return false;
+    };
+  }
 };
 
 /// create a processor spec

--- a/Detectors/ITSMFT/MFT/workflow/src/RecoWorkflow.cxx
+++ b/Detectors/ITSMFT/MFT/workflow/src/RecoWorkflow.cxx
@@ -20,6 +20,7 @@
 #include "ITSMFTWorkflow/DigitReaderSpec.h"
 #include "MFTWorkflow/MFTAssessmentSpec.h"
 #include "MFTWorkflow/TracksToRecordsSpec.h"
+#include "GlobalTrackingWorkflowReaders/IRFrameReaderSpec.h"
 
 namespace o2
 {
@@ -51,6 +52,11 @@ framework::WorkflowSpec getWorkflow(
   if (!disableRootOutput) {
     specs.emplace_back(o2::mft::getClusterWriterSpec(useMC));
   }
+  auto& trackingParam = MFTTrackingParam::Instance();
+  if (trackingParam.irFramesOnly) {
+    specs.emplace_back(o2::globaltracking::getIRFrameReaderSpec("ITS", 0, "its-irframe-reader", "o2_its_irframe.root"));
+  }
+
   if (runTracking) {
     specs.emplace_back(o2::mft::getTrackerSpec(useMC, nThreads));
     if (!disableRootOutput) {

--- a/Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx
+++ b/Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx
@@ -54,7 +54,7 @@ void TrackerDPL::init(InitContext& ic)
   }
 
   // tracking configuration parameters
-  auto& trackingParam = MFTTrackingParam::Instance(); // to avoi loading interpreter during the run
+  auto& trackingParam = MFTTrackingParam::Instance(); // to avoid loading interpreter during the run
 }
 
 void TrackerDPL::run(ProcessingContext& pc)
@@ -73,16 +73,34 @@ void TrackerDPL::run(ProcessingContext& pc)
   auto rofsinput = pc.inputs().get<const std::vector<o2::itsmft::ROFRecord>>("ROframes");
   auto& rofs = pc.outputs().make<std::vector<o2::itsmft::ROFRecord>>(Output{"MFT", "MFTTrackROF", 0, Lifetime::Timeframe}, rofsinput.begin(), rofsinput.end());
 
-  LOG(info) << "MFTTracker pulled " << compClusters.size() << " compressed clusters in "
-            << rofsinput.size() << " RO frames";
+  ROFFilter filter = [](const o2::itsmft::ROFRecord& r) { return true; };
+
+  LOG(info) << "MFTTracker pulled " << compClusters.size() << " compressed clusters in " << rofsinput.size() << " RO frames";
+
+  auto& trackingParam = MFTTrackingParam::Instance();
+  if (trackingParam.irFramesOnly) {
+    // selects only those ROFs that overlap ITS IRFrame
+    LOG(info) << "MFTTracker IRFrame filter enabled: loading ITS IR Frames. ";
+    auto irFrames = pc.inputs().get<gsl::span<o2::dataformats::IRFrame>>("IRFramesITS");
+    filter = createIRFrameFilter(irFrames);
+
+    if (fair::Logger::Logging(fair::Severity::debug)) {
+      for (const auto& irf : irFrames) {
+        LOG(debug) << "IRFrame.info = " << irf.info << " ; min = " << irf.getMin().bc << " ; max = " << irf.getMax().bc;
+      }
+    }
+  }
+
+  if (trackingParam.isMultCutRequested()) {
+    LOG(info) << "MFTTracker multiplicity filter enabled. ROF selection: Min nClusters =  " << trackingParam.cutMultClusLow << " ; Max nClusters = " << trackingParam.cutMultClusHigh;
+  }
 
   const dataformats::MCTruthContainer<MCCompLabel>* labels = mUseMC ? pc.inputs().get<const dataformats::MCTruthContainer<MCCompLabel>*>("labels").release() : nullptr;
   gsl::span<itsmft::MC2ROFRecord const> mc2rofs;
   if (mUseMC) {
     // get the array as read-only span, a snapshot of the object is sent forward
     mc2rofs = pc.inputs().get<gsl::span<itsmft::MC2ROFRecord>>("MC2ROframes");
-    LOG(info) << labels->getIndexedSize() << " MC label objects , in "
-              << mc2rofs.size() << " MC events";
+    LOG(info) << labels->getIndexedSize() << " MC label objects , in " << mc2rofs.size() << " MC events";
   }
 
   auto& allClusIdx = pc.outputs().make<std::vector<int>>(Output{"MFT", "TRACKCLSID", 0, Lifetime::Timeframe});
@@ -98,16 +116,15 @@ void TrackerDPL::run(ProcessingContext& pc)
   LOG(debug) << "nROFs = " << nROFs << " rofsPerWorker = " << rofsPerWorker;
 
   auto loadData = [&, this](auto& trackerVec, auto& roFrameDataVec) {
-    auto& trackingParam = MFTTrackingParam::Instance();
     auto& tracker = trackerVec[0]; // Use first tracker to load the data: serial operation
     gsl::span<const unsigned char>::iterator pattIt = patterns.begin();
 
     auto iROF = 0;
 
-    for (auto& rof : rofs) {
+    for (const auto& rof : rofs) {
       int worker = std::min(int(iROF / rofsPerWorker), mNThreads - 1);
       auto& roFrameData = roFrameDataVec[worker].emplace_back();
-      int nclUsed = ioutils::loadROFrameData(rof, roFrameData, compClusters, pattIt, mDict, labels, tracker.get());
+      int nclUsed = ioutils::loadROFrameData(rof, roFrameData, compClusters, pattIt, mDict, labels, tracker.get(), filter);
       roFrameData.initialize(trackingParam.FullClusterScan);
       LOG(debug) << "ROframeId: " << iROF << ", clusters loaded : " << nclUsed << " on worker " << worker;
       iROF++;
@@ -330,14 +347,20 @@ void TrackerDPL::updateTimeDependentParams(ProcessingContext& pc)
   if (!initOnceDone) { // this params need to be queried only once
     initOnceDone = true;
     pc.inputs().get<o2::itsmft::TopologyDictionary*>("cldict"); // just to trigger the finaliseCCDB
-    // tracking configuration parameters
-
     bool continuous = o2::base::GRPGeomHelper::instance().getGRPECS()->isDetContinuousReadOut(o2::detectors::DetID::MFT);
-    LOG(info) << "MFTTracker RO: continuous=" << continuous;
+    LOG(info) << "MFTTracker RO: continuous =" << continuous;
+    mMFTTriggered = !continuous;
+    const auto& alpParams = o2::itsmft::DPLAlpideParam<o2::detectors::DetID::MFT>::Instance();
+    if (mMFTTriggered) {
+      setMFTROFrameLengthMUS(alpParams.roFrameLengthTrig / 1.e3); // MFT ROFrame duration in \mus
+    } else {
+      setMFTROFrameLengthInBC(alpParams.roFrameLengthInBC); // MFT ROFrame duration in \mus
+    }
 
     o2::mft::GeometryTGeo* geom = o2::mft::GeometryTGeo::Instance();
     geom->fillMatrixCache(o2::math_utils::bit2Mask(o2::math_utils::TransformType::T2L, o2::math_utils::TransformType::T2GRot,
                                                    o2::math_utils::TransformType::T2G));
+    // tracking configuration parameters
     auto& trackingParam = MFTTrackingParam::Instance();
     auto field = static_cast<o2::field::MagneticField*>(TGeoGlobalMagField::Instance()->GetField());
     double centerMFT[3] = {0, 0, -61.4}; // Field at center of MFT
@@ -377,6 +400,23 @@ void TrackerDPL::finaliseCCDB(ConcreteDataMatcher& matcher, void* obj)
   }
 }
 
+///_______________________________________
+void TrackerDPL::setMFTROFrameLengthMUS(float fums)
+{
+  mMFTROFrameLengthMUS = fums;
+  mMFTROFrameLengthMUSInv = 1. / mMFTROFrameLengthMUS;
+  mMFTROFrameLengthInBC = std::max(1, int(mMFTROFrameLengthMUS / (o2::constants::lhc::LHCBunchSpacingNS * 1e-3)));
+}
+
+///_______________________________________
+void TrackerDPL::setMFTROFrameLengthInBC(int nbc)
+{
+  mMFTROFrameLengthInBC = nbc;
+  mMFTROFrameLengthMUS = nbc * o2::constants::lhc::LHCBunchSpacingNS * 1e-3;
+  mMFTROFrameLengthMUSInv = 1. / mMFTROFrameLengthMUS;
+}
+
+///_______________________________________
 DataProcessorSpec getTrackerSpec(bool useMC, int nThreads)
 {
   std::vector<InputSpec> inputs;
@@ -384,6 +424,12 @@ DataProcessorSpec getTrackerSpec(bool useMC, int nThreads)
   inputs.emplace_back("patterns", "MFT", "PATTERNS", 0, Lifetime::Timeframe);
   inputs.emplace_back("ROframes", "MFT", "CLUSTERSROF", 0, Lifetime::Timeframe);
   inputs.emplace_back("cldict", "MFT", "CLUSDICT", 0, Lifetime::Condition, ccdbParamSpec("MFT/Calib/ClusterDictionary"));
+
+  auto& trackingParam = MFTTrackingParam::Instance();
+  if (trackingParam.irFramesOnly) {
+    inputs.emplace_back("IRFramesITS", "ITS", "IRFRAMES", 0, Lifetime::Timeframe);
+  }
+
   auto ggRequest = std::make_shared<o2::base::GRPGeomRequest>(false,                             // orbitResetTime
                                                               true,                              // GRPECS=true
                                                               false,                             // GRPLHCIF

--- a/prodtests/full-system-test/dpl-workflow.sh
+++ b/prodtests/full-system-test/dpl-workflow.sh
@@ -64,6 +64,7 @@ TOF_CONFIG=
 TOF_INPUT=raw
 TOF_OUTPUT=clusters
 ITS_CONFIG_KEY=
+MFT_CONFIG_KEY=
 TRD_CONFIG=
 TRD_CONFIG_KEY=
 TRD_FILTER_CONFIG=
@@ -86,6 +87,11 @@ if [[ $SYNCMODE == 1 ]]; then
     ITS_CONFIG_KEY+="fastMultConfig.cutMultClusLow=30;fastMultConfig.cutMultClusHigh=2000;fastMultConfig.cutMultVtxHigh=500;"
   elif [[ $BEAMTYPE == "pp" ]]; then
     ITS_CONFIG_KEY+="fastMultConfig.cutMultClusLow=-1;fastMultConfig.cutMultClusHigh=-1;fastMultConfig.cutMultVtxHigh=-1;ITSVertexerParam.phiCut=0.5;ITSVertexerParam.clusterContributorsCut=3;ITSVertexerParam.tanLambdaCut=0.2;fastMultConfig.cutRandomFraction=0.9;"
+  fi
+  if has_detector_reco ITS; then
+    MFT_CONFIG_KEY+="MFTTracking.irFramesOnly=1;"
+  else
+    MFT_CONFIG_KEY+="MFTTracking.cutMultClusLow=0;MFTTracking.cutMultClusHigh=2000;"
   fi
 
   PVERTEXING_CONFIG_KEY+="pvertexer.meanVertexExtraErrConstraint=0.3;" # for calibration relax the constraint
@@ -286,7 +292,7 @@ if has_processing_step MUON_SYNC_RECO; then
     fi
     has_detector_reco ITS && [[ $RUNTYPE != "COSMICS" ]] && CONFIG_EXTRA_PROCESS_o2_mch_reco_workflow+="MCHTimeClusterizer.irFramesOnly=true;"
   fi
-  [[ $RUNTYPE == "COSMICS" ]] && [[ -z $CONFIG_EXTRA_PROCESS_o2_mft_reco_workflow ]] && CONFIG_EXTRA_PROCESS_o2_mft_reco_workflow="MFTTracking.LTFclsRCut=0.2;MFTTracking.forceZeroField=true;MFTTracking.FullClusterScan=true"
+  [[ $RUNTYPE == "COSMICS" ]] && [[ -z $CONFIG_EXTRA_PROCESS_o2_mft_reco_workflow ]] && CONFIG_EXTRA_PROCESS_o2_mft_reco_workflow="MFTTracking.FullClusterScan=true"
 fi
 [[ "0$ED_VERTEX_MODE" == "01" ]] && has_detectors_reco ITS && has_detector_matching PRIMVTX && [[ ! -z "$VERTEXING_SOURCES" ]] && EVE_CONFIG+=" --primary-vertex-mode"
 [[ $EPNSYNCMODE == 1 ]] && [[ -z $CONFIG_EXTRA_PROCESS_o2_trd_global_tracking ]] && CONFIG_EXTRA_PROCESS_o2_trd_global_tracking='GPU_rec_trd.maxChi2=25;GPU_rec_trd.penaltyChi2=20;GPU_rec_trd.extraRoadY=4;GPU_rec_trd.extraRoadZ=10;GPU_rec_trd.applyDeflectionCut=0;GPU_rec_trd.trkltResRPhiIdeal=1'
@@ -413,7 +419,7 @@ has_detectors TPC && [ -z "$DISABLE_ROOT_OUTPUT" ] && add_W o2-tpc-reco-workflow
 # Reconstruction workflows normally active only in async mode in async mode ($LIST_OF_ASYNC_RECO_STEPS), but can be forced via $WORKFLOW_EXTRA_PROCESSING_STEPS
 has_detector MID && has_processing_step MID_RECO && add_W o2-mid-reco-workflow "$DISABLE_ROOT_OUTPUT $DISABLE_MC --pipeline $(get_N MIDClusterizer MID REST 1),$(get_N MIDTracker MID REST 1)"
 has_detector MCH && has_processing_step MCH_RECO && add_W o2-mch-reco-workflow "$DISABLE_DIGIT_ROOT_INPUT $DISABLE_ROOT_OUTPUT $DISABLE_MC --pipeline $(get_N mch-track-finder MCH REST 1 MCHTRK),$(get_N mch-cluster-finder MCH REST 1 MCHCL),$(get_N mch-cluster-transformer MCH REST 1)" "$MCH_CONFIG_KEY"
-has_detector MFT && has_processing_step MFT_RECO && add_W o2-mft-reco-workflow "$DISABLE_DIGIT_CLUSTER_INPUT $DISABLE_MC $DISABLE_ROOT_OUTPUT --pipeline $(get_N mft-tracker MFT REST 1 MFTTRK)" "$ITSMFT_STROBES"
+has_detector MFT && has_processing_step MFT_RECO && add_W o2-mft-reco-workflow "$DISABLE_DIGIT_CLUSTER_INPUT $DISABLE_MC $DISABLE_ROOT_OUTPUT --pipeline $(get_N mft-tracker MFT REST 1 MFTTRK)" "$MFT_CONFIG_KEY;$ITSMFT_STROBES"
 has_detector FDD && has_processing_step FDD_RECO && add_W o2-fdd-reco-workflow "$DISABLE_DIGIT_ROOT_INPUT $DISABLE_ROOT_OUTPUT $DISABLE_MC"
 has_detector FV0 && has_processing_step FV0_RECO && add_W o2-fv0-reco-workflow "$DISABLE_DIGIT_ROOT_INPUT $DISABLE_ROOT_OUTPUT $DISABLE_MC"
 has_detector ZDC && has_processing_step ZDC_RECO && add_W o2-zdc-digits-reco "$DISABLE_DIGIT_ROOT_INPUT $DISABLE_ROOT_OUTPUT $DISABLE_MC"

